File: crates/wem-analysis/src/psychoacoustics/temporal.rs
//! Temporal weighting and history relaxation for psychoacoustic analysis.
//!
//! Mirrors Python `wwise_wem/analysis/psychoacoustics/temporal.py`. Float32
//! storage boundaries are `f32_of`; the width table is reconstructed exactly.

use crate::config::{f32_of, AnalysisError};

/// Sliding relaxation widths for the 128-bin short-block history recurrence
/// (Python `SHORT_HISTORY_RELAXATION_WIDTHS`). The repeated pushes are
/// intentional: they reconstruct the frozen width table exactly.
#[allow(clippy::same_item_push)]
pub fn short_history_relaxation_widths() -> Vec<i64> {
    let mut widths: Vec<i64> = Vec::with_capacity(128);
    for _ in 0..8 {
        widths.push(1);
    }
    for width in 2..25 {
        for _ in 0..4 {
            widths.push(width);
        }
    }
    for _ in 0..3 {
        widths.push(25);
    }
    for width in (1..=24).rev() {
        widths.push(width);
    }
    widths.push(1);
    debug_assert_eq!(widths.len(), 128);
    widths
}

/// Inputs to the temporal kernel (Python `TemporalKernelInputs`).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct TemporalKernelInputs {
    pub bins: i64,
    pub run_count: i64,
    pub tail_count: i64,
    pub high_rate: bool,
    pub curve_bias: f64,
    pub transition_code: i64,
    pub previous_transition: i64,
    pub update_gate: i64,
    pub analysis_mode: i64,
}

/// Result of the temporal kernel (Python `TemporalKernelResult`).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct TemporalKernelResult {
    pub active: i64,
    pub update_state: i64,
    pub lower_weight: f64,
    pub upper_weight: f64,
    pub state_bias: f64,
    pub chase_seed: f64,
}

impl Default for TemporalKernelResult {
    fn default() -> Self {
        Self {
            active: 0,
            update_state: 0,
            lower_weight: 0.0,
            upper_weight: 0.0,
            state_bias: 0.0,
            chase_seed: 0.0,
        }
    }
}

impl TemporalKernelResult {
    /// The packed word tuple (Python `words`).
    pub fn words(&self) -> (i64, i64, f64, f64, f64, f64) {
        (
            self.active,
            self.update_state,
            self.lower_weight,
            self.upper_weight,
            self.state_bias,
            self.chase_seed,
        )
    }
}

/// Compute temporal weights for the active block geometry and transition
/// (Python `compute_temporal_kernel`).
pub fn compute_temporal_kernel(inputs: &TemporalKernelInputs) -> TemporalKernelResult {
    if !inputs.high_rate {
        return TemporalKernelResult::default();
    }

    let update = if inputs.update_gate == 0 || inputs.analysis_mode == 2 {
        1
    } else {
        0
    };
    if update == 0 && inputs.analysis_mode == 0 {
        return TemporalKernelResult::default();
    }
    if inputs.transition_code != 0 {
        return TemporalKernelResult {
            update_state: update,
            ..TemporalKernelResult::default()
        };
    }

    if inputs.bins == 128 {
        let multiplier = if inputs.curve_bias >= 3.0 { 3 } else { 2 };
        let result = if inputs.previous_transition != 0 {
            TemporalKernelResult {
                active: 1,
                update_state: update,
                lower_weight: 0.7,
                upper_weight: 0.0,
                state_bias: 0.0,
                chase_seed: 8.0,
            }
        } else if inputs.run_count >= 8 {
            TemporalKernelResult {
                active: 1,
                update_state: update,
                lower_weight: 0.3,
                upper_weight: 0.0,
                state_bias: 25.0,
                chase_seed: 0.0,
            }
        } else {
            TemporalKernelResult {
                active: 1,
                update_state: update,
                lower_weight: 0.7 - (inputs.run_count - 1) as f64 / 17.0,
                upper_weight: 0.0,
                state_bias: (inputs.run_count * multiplier) as f64,
                chase_seed: (8 - inputs.run_count) as f64,
            }
        };
        let divisor = 8.0;

        let mut lower = result.lower_weight;
        if inputs.tail_count != 0 {
            lower *= inputs.tail_count as f64 / divisor;
        }
        if inputs.update_gate != 0 && inputs.analysis_mode == 0 {
            lower *= 0.2;
        }
        TemporalKernelResult {
            active: result.active,
            update_state: result.update_state,
            lower_weight: lower,
            upper_weight: result.upper_weight,
            state_bias: result.state_bias,
            chase_seed: result.chase_seed,
        }
    } else if inputs.bins == 256 {
        if inputs.previous_transition != 0 {
            TemporalKernelResult {
                active: 1,
                update_state: update,
                lower_weight: 0.6,
                upper_weight: 0.0,
                state_bias: 12.0,
                chase_seed: 8.0,
            }
        } else if inputs.run_count >= 4 {
            TemporalKernelResult {
                active: 1,
                update_state: update,
                lower_weight: 0.2,
                upper_weight: 0.0,
                state_bias: 30.0,
                chase_seed: 0.0,
            }
        } else {
            let lower_weight = 0.4 - (inputs.run_count - 1) as f64 / 11.0;
            let state_bias = 2.0 * (3.0 * inputs.run_count as f64 + 6.0);
            let chase_seed = 2.0 * (4.0 - inputs.run_count as f64);
            TemporalKernelResult {
                active: 1,
                update_state: update,
                lower_weight,
                upper_weight: 0.0,
                state_bias,
                chase_seed,
            }
        }
    } else {
        // Long blocks do not run the short peak-continuation branch.
        TemporalKernelResult {
            update_state: update,
            ..TemporalKernelResult::default()
        }
    }
}

/// Apply the history rebase that precedes width relaxation
/// (Python `rebase_history`).
pub fn rebase_history(
    inputs: &TemporalKernelInputs,
    result: &TemporalKernelResult,
    state: &[f64],
    history: &[f64],
) -> Result<Vec<f64>, AnalysisError> {
    if state.len() as i64 != inputs.bins || history.len() as i64 != inputs.bins {
        return Err(AnalysisError::RebaseLengthMismatch {
            want: inputs.bins,
        });
    }
    if result.active == 0 || result.update_state == 0 {
        return Ok(history.to_vec());
    }
    let delta = if inputs.bins == 128 {
        5.0
    } else if inputs.bins == 256 {
        10.0
    } else {
        return Ok(history.to_vec());
    };
    let source = if inputs.previous_transition != 0 { state } else { history };
    Ok(source.iter().map(|value| f32_of(*value - delta)).collect())
}

/// Apply sliding width-table relaxation to one history curve
/// (Python `relax_history`).
pub fn relax_history(
    history: &[f64],
    raw: &[f64],
    widths: &[i64],
    delta: f64,
) -> Result<Vec<f64>, AnalysisError> {
    if !(history.len() == raw.len() && history.len() == widths.len()) {
        return Err(AnalysisError::RelaxLengthMismatch);
    }
    let mut out: Vec<f64> = history.iter().map(|value| f32_of(*value)).collect();
    let n = out.len();
    for (k, &width) in widths.iter().enumerate() {
        if width <= 1 {
            continue;
        }
        for j in 1..width {
            let index = k + j as usize;
            if index >= n {
                break;
            }
            if raw[k] - j as f64 * 75.0 / width as f64 > out[index] {
                let denominator = widths[index];
                if denominator <= 0 {
                    return Err(AnalysisError::RelaxWidthNonPositive);
                }
                // Each relaxation is stored as float32 before later comparisons.
                out[index] = f32_of(out[index] + delta / denominator as f64);
            }
        }
    }
    Ok(out)
}

/// Apply the complete 128-bin short-history relaxation table
/// (Python `relax_short_history`).
pub fn relax_short_history(history: &[f64], raw: &[f64]) -> Result<Vec<f64>, AnalysisError> {
    relax_history(history, raw, &short_history_relaxation_widths(), 5.0)
}

/// Update short-block temporal history from local psychoacoustic surfaces
/// (Python `update_short_history`).
#[allow(clippy::too_many_arguments)]
pub fn update_short_history(
    inputs: &TemporalKernelInputs,
    state: &[f64],
    history: &[f64],
    raw: &[f64],
    seed: &[f64],
    remap: &[f64],
    mask_curve: &[f64],
    curve_cap: f64,
    q: f64,
    candidate_bound: i64,
) -> Result<Vec<f64>, AnalysisError> {
    if inputs.bins != 128 {
        return Err(AnalysisError::ShortTemporalBins { want: 128 });
    }
    let arrays = [state, history, raw, seed, remap, mask_curve];
    if arrays.iter().any(|values| values.len() != 128) {
        return Err(AnalysisError::ShortTemporalLength { want: 128 });
    }
    let result = compute_temporal_kernel(inputs);
    let mut baseline = rebase_history(inputs, &result, state, history)?;
    if result.active != 0 && result.update_state != 0 {
        baseline = relax_short_history(&baseline, raw)?;
    } else {
        for value in baseline.iter_mut() {
            *value = f32_of(*value);
        }
    }

    if result.active == 0 {
        return Ok(baseline);
    }
    let mut out = baseline.clone();
    let mut subtract = 0.0;
    if q >= 0.0 && inputs.curve_bias >= 25.0 {
        subtract = q * (inputs.curve_bias - 25.0);
    }
    for index in 0..128usize {
        let cap = f32_of(mask_curve[index] + remap[index]).min(curve_cap);
        let mut candidate = f32_of(seed[index] + inputs.curve_bias);
        if index as i64 <= candidate_bound {
            candidate = f32_of(candidate - subtract);
        }
        if candidate < cap
            && state[index] < cap
            && out[index] + result.state_bias < raw[index]
            && result.update_state != 0 {
                out[index] = f32_of(raw[index]);
            }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn width_table_length() {
        assert_eq!(short_history_relaxation_widths().len(), 128);
    }

    #[test]
    fn kernel_low_rate_defaults() {
        let inputs = TemporalKernelInputs {
            bins: 128,
            run_count: 1,
            tail_count: 0,
            high_rate: false,
            curve_bias: 0.0,
            transition_code: 0,
            previous_transition: 0,
            update_gate: 0,
            analysis_mode: 0,
        };
        let result = compute_temporal_kernel(&inputs);
        assert_eq!(result.active, 0);
    }
}

File: reference/wwise_wem_reference/analysis/psychoacoustics/temporal.py
#!/usr/bin/env python3
"""Temporal weighting and history relaxation for psychoacoustic analysis."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Sequence

from ..._f32 import _f32


# Sliding relaxation widths for the 128-bin short-block history recurrence.
SHORT_HISTORY_RELAXATION_WIDTHS = (
    (1,) * 8
    + tuple(width for width in range(2, 25) for _ in range(4))
    + (25,) * 3
    + tuple(range(24, 0, -1))
    + (1,)
)
if len(SHORT_HISTORY_RELAXATION_WIDTHS) != 128:
    raise AssertionError("short history width table has the wrong length")


@dataclass(frozen=True)
class TemporalKernelInputs:
    bins: int
    run_count: int
    tail_count: int
    high_rate: bool
    curve_bias: float
    transition_code: int
    previous_transition: int
    hold_update: int
    analysis_mode: int


@dataclass(frozen=True)
class TemporalKernelResult:
    active: int = 0
    update_state: int = 0
    lower_weight: float = 0.0
    upper_weight: float = 0.0
    state_bias: float = 0.0
    chase_seed: float = 0.0

    def words(self) -> tuple[int | float, ...]:
        return (
            self.active,
            self.update_state,
            self.lower_weight,
            self.upper_weight,
            self.state_bias,
            self.chase_seed,
        )


def compute_temporal_kernel(inputs: TemporalKernelInputs) -> TemporalKernelResult:
    """Compute temporal weights for the active block geometry and transition."""
    if not inputs.high_rate:
        return TemporalKernelResult()

    update = int(not inputs.hold_update or inputs.analysis_mode == 2)
    if not update and inputs.analysis_mode == 0:
        return TemporalKernelResult()
    if inputs.transition_code:
        return TemporalKernelResult(update_state=update)

    if inputs.bins == 128:
        multiplier = 3 if inputs.curve_bias >= 3.0 else 2
        if inputs.previous_transition:
            result = TemporalKernelResult(1, update, 0.7, 0.0, 0.0, 8.0)
        elif inputs.run_count >= 8:
            result = TemporalKernelResult(
                1, update, 0.3, 0.0, 25.0, 0.0
            )
        else:
            result = TemporalKernelResult(
                1,
                update,
                0.7 - (inputs.run_count - 1) / 17.0,
                0.0,
                float(inputs.run_count * multiplier),
                float(8 - inputs.run_count),
            )
        divisor = 8.0
    elif inputs.bins == 256:
        if inputs.previous_transition:
            result = TemporalKernelResult(1, update, 0.6, 0.0, 12.0, 8.0)
        elif inputs.run_count >= 4:
            result = TemporalKernelResult(1, update, 0.2, 0.0, 30.0, 0.0)
        else:
            result = TemporalKernelResult(
                1,
                update,
                0.4 - (inputs.run_count - 1) / 11.0,
                0.0,
                float(2 * (3 * inputs.run_count + 6)),
                float(2 * (4 - inputs.run_count)),
            )
        divisor = 16.0
    else:
        # Long blocks do not run the short peak-continuation branch.
        return TemporalKernelResult(update_state=update)

    lower = result.lower_weight
    if inputs.tail_count:
        lower *= inputs.tail_count / divisor
    if inputs.hold_update and inputs.analysis_mode == 0:
        lower *= 0.2
    return TemporalKernelResult(
        result.active,
        result.update_state,
        lower,
        result.upper_weight,
        result.state_bias,
        result.chase_seed,
    )


def rebase_history(
    inputs: TemporalKernelInputs,
    result: TemporalKernelResult,
    state: Sequence[float],
    history: Sequence[float],
) -> list[float]:
    """Apply the history rebase that precedes width relaxation.

    The caller can then apply :func:`relax_history` with the static width table
    when an n=128/256 path needs it.  n=1024, all transitions, and inactive
    tuples leave history unchanged.
    """
    if len(state) != inputs.bins or len(history) != inputs.bins:
        raise ValueError("state/history length must equal n")
    if not result.active or not result.update_state:
        return [float(value) for value in history]
    if inputs.bins == 128:
        delta = 5.0
    elif inputs.bins == 256:
        delta = 10.0
    else:
        return [float(value) for value in history]
    source = state if inputs.previous_transition else history
    return [_f32(float(value) - delta) for value in source]


def relax_history(
    history: Sequence[float],
    raw: Sequence[float],
    widths: Sequence[int],
    *,
    delta: float,
) -> list[float]:
    """Apply sliding width-table relaxation to one history curve."""
    if not (len(history) == len(raw) == len(widths)):
        raise ValueError("history/raw/widths must have equal length")
    out = [_f32(value) for value in history]
    n = len(out)
    for k, width in enumerate(widths):
        if width <= 1:
            continue
        width = int(width)
        for j in range(1, width):
            index = k + j
            if index >= n:
                break
            if float(raw[k]) - j * 75.0 / width > out[index]:
                denominator = int(widths[index])
                if denominator <= 0:
                    raise ValueError("history width table contains a nonpositive value")
                # Each relaxation is stored as float32 before later comparisons.
                out[index] = _f32(out[index] + delta / denominator)
    return out


def relax_short_history(
    history: Sequence[float], raw: Sequence[float]
) -> list[float]:
    """Apply the complete 128-bin short-history relaxation table."""
    return relax_history(
        history, raw, SHORT_HISTORY_RELAXATION_WIDTHS, delta=5.0
    )


def update_short_history(
    inputs: TemporalKernelInputs,
    state: Sequence[float],
    history: Sequence[float],
    raw: Sequence[float],
    seed: Sequence[float],
    remap: Sequence[float],
    mask_curve: Sequence[float],
    *,
    curve_cap: float,
    q: float = 0.0,
    candidate_bound: int = 0,
) -> list[float]:
    """Update short-block temporal history from local psychoacoustic surfaces."""
    if inputs.bins != 128:
        raise ValueError("short temporal history expects 128 bins")
    arrays = (state, history, raw, seed, remap, mask_curve)
    if any(len(values) != 128 for values in arrays):
        raise ValueError("short temporal history curves must have 128 bins")
    result = compute_temporal_kernel(inputs)
    baseline = rebase_history(inputs, result, state, history)
    if result.active and result.update_state:
        baseline = relax_short_history(baseline, raw)
    else:
        baseline = [_f32(value) for value in baseline]

    if not result.active:
        return baseline
    out = baseline[:]
    subtract = 0.0
    if float(q) >= 0.0 and inputs.curve_bias >= 25.0:
        subtract = float(q) * (inputs.curve_bias - 25.0)
    for index in range(128):
        cap = min(
            _f32(float(mask_curve[index]) + float(remap[index])),
            float(curve_cap),
        )
        candidate = _f32(float(seed[index]) + inputs.curve_bias)
        if index <= candidate_bound:
            candidate = _f32(candidate - subtract)
        if (
            candidate < cap
            and float(state[index]) < cap
            and float(out[index]) + result.state_bias < float(raw[index])
        ):
            if result.update_state:
                out[index] = _f32(raw[index])
    return out
